src/snowcli/cli/nativeapp/utils.py

```diff
@@ -1,15 +1,29 @@
-from typing import Callable, Optional
+from functools import cache
+from sys import stdin, stdout
+from typing import Callable, List, Optional
 
 from snowflake.connector.cursor import DictCursor
 
 
-def find_row(cursor: DictCursor, predicate: Callable[[dict], bool]) -> Optional[dict]:
+def _rows_generator(cursor: DictCursor, predicate: Callable[[dict], bool]):
+    return (row for row in cursor.fetchall() if predicate(row))
+
+
+def find_all_rows(cursor: DictCursor, predicate: Callable[[dict], bool]) -> List[dict]:
+    return list(_rows_generator(cursor, predicate))
+
+
+def find_first_row(
+    cursor: DictCursor, predicate: Callable[[dict], bool]
+) -> Optional[dict]:
     """Returns the first row that matches the predicate, or None."""
-    return next(
-        (row for row in cursor.fetchall() if predicate(row)),
-        None,
-    )
+    return next(_rows_generator(cursor, predicate), None)
 
 
 def needs_confirmation(needs_confirm: bool, auto_yes: bool) -> bool:
     return needs_confirm and not auto_yes
+
+
+@cache
+def is_user_in_interactive_mode():
```

Comment:
Feels like this isn't in the right place. It's not uncommon to provide CLI options to force one way or the other (--non-interactive/--force-interactive). IMHO it would be a lot easier to bundle this with policy selection at the entry point, e.g.:

```
if force:
    policy = AllowAll()
elif is_interactive():
    policy = AlwaysAsk()
else:
    policy = DenyAll()
```

It would then be trivial to introduce flags to change the default behaviour in the future, if needed.

Reply:
deny by default is odd in an interactive shell, we should ask instead. This is how most dev tools behave.

To be clear, I wasn't asking to include interactive/non-interactive switches, I was just pointing out that the implementation would make those difficult to add later. Usually this, to me, suggests that responsibilities might need to be moved around a bit, which is what I suggested.